onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -115,6 +115,11 @@ Status Attention<T>::Compute(OpKernelContext* context) const {
   TensorShape output_shape(dims);
   Tensor* output = context->Output(0, output_shape);
 
+  // If the given batch of sequences is empty, stop processing right here
```

Comment:
It is better to verify input shape size in AttentionBase::CheckInputs instead. Then you need not add such check in every class that derived from AttentionBase (for example QAttention) .

Reply:
This is not a check. This is to return early and not do downstream processing if the output is going to be empty (if sequence length is 0 for example). This logic is going to be hard to add into a base class right ?

Reply:
Added the early return logic to the quantized versions of these ops as well. As stated above, this logic has to be in Compute() (and hence cannot be put in the base class)

Reply:
CheckInputs are called at the begining of Compute(), so CheckInputs can also early return (a failure status) by checking empty input shape. Note that the output shape is same as input shape.

Reply:
It is true that CheckInputs() can indicate to the caller that there is no op processing required. But as you said, we would have to return something other than Status::OK() - essentially a "failure" that the caller needs to understand is not a "true" failure but rather a failure to indicate that no op processing is required (and hence the caller will return Status:OK() at that point). It seems a little too hacky just to avoid a couple of lines of code in Compute(). I think I will just keep it like this.

Reply:
You need the call to create the output with the correct dimensions too don't you? Even though one or more dims may have a value of zero, the rank is still important. As such I would have expected you need to do this in each Compute. 

---
In reply to: [436246135](https://github.com/microsoft/onnxruntime/pull/4129#discussion_r436246135) [](ancestors = 436246135)